In this exercise, we will:
1) Retrieve data from shared Data Libraries and import to a history
2) Organize RNA-seq data into collections
3) Tag these (two) collections with descriptive tags
4) Run FastQC on these data to get an idea of the quality of these samples

Remember to use the Computational Genomics Course [Galaxy instance](http://ecg2018.bioch.virginia.edu/galaxy/).


Getting data from Data Libraries
Data libraries are a shared location in Galaxy for storing data and are frequently used to make data available. 
We will be using data from the RNA-seq tutorial that will be used in Friday afternoon's workshop. 

_The data provided here are part of a Galaxy tutorial that analyzes RNA-seq data from a study published by [Wu et al. in 2014](DOI:10.1101/gr.164830.113). The goal of this study was to investigate “the dynamics of occupancy and the role in gene regulation of the transcription factor Tal1, a critical regulator of hematopoiesis, at multiple stages of hematopoietic differentiation.” To this end, RNA-seq libraries were constructed from multiple mouse cell types including G1E - a GATA-null immortalized cell line derived from targeted disruption of GATA-1 in mouse embryonic stem cells - and megakaryocytes. This RNA-seq data was used to determine differential gene expression between G1E and megakaryocytes and later correlated with Tal1 occupancy. This dataset (GEO Accession: GSE51338) consists of two biological replicates, paired-end, and poly(A) selected RNA-seq libraries. Because of the long processing time for the large original files, we have downsampled the original raw data files to include only reads that align to chromosome 19 and a subset of interesting genomic loci identified by Wu et al._

In Galaxy, go to _Shared Data -> Data Libraries_ and click through to the "RNA-seq" data. From here, select the eight samples corresponding to the G1E or megakaryocyte cellular stages and import them to a new history.  



The introductions to dataset collections and tags in Galaxy provide the necessary background to compelte this exercise. Feel free to explore the Galaxy interface and ways to organize your data. 

In Galaxy you perform data analyses and organize your data simply by clicking on things. If you have just a few items in your history, clicking is easy. However, in most real-world analyses you never have just a few datasets, instead you have many (sometimes thousands) and **Collections** help manage your data to minimize the amount of clicking you have to do. 

|      |
|------|
|![](/src/tutorials/collections/smallVsLarge.png)|
|<small>**Figure 1. A history with few datasets** is *easy* to navigate. **A history with many datasets** is *hard* to navigate.</small>

Not only may clicking on thousands datasets lead to a severe form of [carpal tunnel syndrome](https://en.wikipedia.org/wiki/Carpal_tunnel_syndrome), it may simply be impossible. **Collections** (also known as *Dataset collections* or *Lists*) described here help to resolve this situation and make very large analyses Galaxy-friendly. 

# Collections combine datasets to minimize clutter

A typical multi-dataset analysis of next-generation sequencing (NGS) data usually involves a large number of sequence datasets, such as, for example, fastq datasets generated by an Illumina machine or downloaded from a Short Read Archive. So it's usually a *collection* of similar things that need to be processed in a *collective* fashion:

|      |
|------|
|![](/src/tutorials/collections/simpleCollection.png)|
|<small>**Figure 2. A collection** is any number of datasets bundled as a **single** entity. In this example, to, say, map *N* fastq files against the human genome you need to manually start *N* mapping jobs. Yet if you first combine individual datasets into a collection, you will only need start a mapping job **once** using the collection as the only input.</small>|

## Collections reflect data complexity

The previous image shows how multiple fastq datasets can be combined in a single collection. But what if the sequencing data one wants to analyze is from a paired-end (or mate-pair) run where each individual sample is represented by *two* fastq files: forward and reverse? Galaxy collections can accommodate this structure:

|      |
|------|
|![](/src/tutorials/collections/pairedCollection.png)|
|<small>**Figure 3. Paired collection** preserves the relationship between read pairs and their respective samples.</small>

# Creating collections in practice

Below we explain how collections can be used in Galaxy. We start with an example where datasets are already in your history. Later in this tutorial we explain how to upload data into collections directly.

## Creating a collection from datasets in your history

As was mentioned above in Galaxy's language there are two interchangeable terms: *a collection* and *a list*. This is because the simplest collection is just a list of datasets.

### A simple collection example

In this example we have six datasets representing single-end (not paired-end) sequencing datasets from six patients. The datasets are already loaded in the history. 

|      |
|------|
|![](/src/tutorials/collections/creating_list.png)|
|<small>**Figure 4. Creating a list (collection) or datasets** (To see a higher resolution image right-click [here](/src/tutorials/collections/creating_list.pdf)). **A**. Click the checkbox <i class="far fa-check-square" aria-hidden="true"></i> icon. This will reveal checkboxes to the left of all datasets in the history. **B**. In this case we want to select all datasets, so press "All" button (alternatively datasets can be filtered as shown [here](/tutorials/histories/#searching-for-datasets)). This will put a check mark into all checkboxes. **C**. Click "For all selected..." button. This will reveal a dropdown. **D**. Since this is **not** paired-end (or mate-pair) data we will choose to "Build Dataset List". This will open a dataset collection creator interface. **E**. Within the dataset collection creator interface use the "Name" box to name the collection. "Hide original elements" checkbox ensures that upon creating the collection the original datasets will be hidden from the history as shown in the next figure. Click "Create list". **F**. A collection named "patients" is now added to the history and original datasets are hidden, so that the history only has one item. **G**. Clicking on collection reveals its content.</small>|

### A paired collection example

In this example we have data from three patients. Each is sequenced in paired-end configuration. This means that there are 3 x 2 = 6 datasets. Just like in the previous example these data are already pre-loaded into history. Note that the datasets are named in the following way:

| Name         | Description                   |
|--------------|-------------------------------|
| `patient1-f` | Forward dataset for patient 1 |
| `patient1-r` | Reverse dataset for patient 1 |
| `patient2-f` | Forward dataset for patient 2 |
| `patient2-r` | Reverse dataset for patient 2 |
| `patient3-f` | Forward dataset for patient 3 |
| `patient3-r` | Reverse dataset for patient 3 |

This is important because the fact that `-f` and `-r` differentiate forward and reverse datasets will be used by collection creator to organize them into pairs (panel **B** of the figure below). Obviously, there may be other ways forward and reverse read datasets are identified such as `-1` and `-2` or `forward` and `reverse` or any other way. These can be used in exactly the same way we use `-f` and `-r` here.

|      |
|------|
|![](/src/tutorials/collections/creating_paired_list.png)|
|<small>**Figure 5. Creating a list (collection) or datasets** (To see a higher resolution image right-click [here](/src/tutorials/collections/creating_paired_list.pdf)). The starting steps are identical to steps A through C of the previous example (Figure 4). **A**. Here choose "Build List of Dataset pairs" to open dataset collection creator. **B**. In the left text box (highlighted in red) enter `-f`. These two characters distinguish *forward* datasets in our case. Once this is done matching datasets will appear underneath the text field. In the right text box (highlighted in green) enter `-r`. These two characters distinguish *reverse* datasets in our case. Again, matching datasets will appear under the box. Finally, click "Auto-pair". **C**. Auto-pairing will form dataset pairs. At this point scroll to the bottom of the form, name the collection, and click "Create list". **D**. A new collection will appear in your history. Clicking the collection will reveal its content. **E**. In this example the collection consists of three pairs. **F**. Clicking on any pair will reveal forward and reverse sequence datasets. </small>|



# Manipulating collections

Collections can renamed, tagged, and manipulated in a number of ways described below.

## Renaming collections

To rename a collection click on its name:

|      |
|------|
|![](/src/tutorials/collections/renaming_collection.png)|
|<small>**Figure 6. Renaming a collection.** **A**. Click on the collection. **B**. Once the collection expands, click on its name. **C**. Enter a new name and hit Enter. **D**. Go back to history (click "Back to..." link). You may need to refresh the history by clicking on the refresh <i class="fas fa-sync" aria-hidden="true"></i> icon.</small>|

## Tagging collections

There are two types of tags that can be used as an additional level of labeling for collections: **standard tags** and **hashtags** (also known as **propagating tags**). **Standard tags** add another level of description to collections making them easier to find. **Hashtags** are much more powerful as they are **displayed** in the history panel and **propagate** through the analysis:

|      |
|------|
|![](/src/tutorials/collections/tagging_collection.png)|
|<small>**Figure 7. Tagging a collection**. **A**. Click on collection. **B**. Once collection expands, type tag name in the tagging box. **If you want the tag to propagate through the analysis add the hash (`#`) symbol in front of the tag as shown here.** **C**. Hit Enter and go back to history by clicking the "Back to..." link. **D**. The collection is now tagged.</small>|

The following video shows the power of hashtags:

<div class="embed-responsive embed-responsive-16by9"><iframe src="https://player.vimeo.com/video/216895965?portrait=0" webkitallowfullscreen mozallowfullscreen allowfullscreen></iframe></div>



# Using collections - a QC example

Earlier in this tutorial we demonstrated what collections are, how they can be created, and manipulated. Now let's have a short usage example. 

Here, we want to have a program called [FastQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/) perform summary QC on our data. 

How do the data look? Do any samples stand out? 



## Using dataset collection tools (additional reading)

Galaxy contains a special set of tools - **Collection operations** - designed for transforming existing collections. The following figure highlights the possibilities:

|      |
|------|
|![](/src/tutorials/collections/collection_operations.png)|
|<small>**Figure 8. Collection operations**. **Zip** - combines two collections as a zipper. If you, for example, have two collections representing forward and reverse reads, respectively, this tool can be used to combine them into a single paired collection. **Unzip** - performs opposite of Zip. **Filter Failed** - when running analyses on collections it is possible that some individual analyses on collection elements will fail. This tool allows you to filter out failed elements so you can continue your analysis instead of performing it again. **Flatten** - squish collection into a simple list of elements. **Filter from File** - given a file of collection element IDs reduce the collection to only a subset of elements listed in the file. **Relabel from File** - use labels in the file to change the collection element names. **Concatenate** - merge two collection tail-to-head.</small>|
# The future

The future improvements of dataset collections will allow representing complex experiments such as the ones shown in the two previous figures in a single collection.

# If things don't work...
...you need to complain. Use [Galaxy's BioStar Channel](https://usegalaxy.org/biostar/biostar_redirect) to do this.
